src/qibo/backends/clifford_operations.py

```diff
@@ -0,0 +1,450 @@
+        symplectic_matrix = CZ(symplectic_matrix, control_q, target_q, nqubits)
+        return X(symplectic_matrix, target_q, nqubits)
+    # theta = 3 * pi + 4 * n * pi
+    elif (theta / np.pi - 3) % 4 == 0:
+        symplectic_matrix = CNOT(symplectic_matrix, control_q, target_q, nqubits)
+        symplectic_matrix = X(symplectic_matrix, target_q, nqubits)
+        symplectic_matrix = CY(symplectic_matrix, control_q, target_q, nqubits)
+        return X(symplectic_matrix, target_q, nqubits)
+
+
+def CRY(symplectic_matrix, control_q, target_q, nqubits, theta):
+    # theta = 4 * n * pi
+    if theta % (4 * np.pi) == 0:
+        return I(symplectic_matrix, target_q, nqubits)
+    # theta = pi + 4 * n * pi
+    elif (theta / np.pi - 1) % 4 == 0:
+        symplectic_matrix = Z(symplectic_matrix, target_q, nqubits)
+        symplectic_matrix = CNOT(symplectic_matrix, control_q, target_q, nqubits)
+        symplectic_matrix = Z(symplectic_matrix, target_q, nqubits)
+        return CZ(symplectic_matrix, control_q, target_q, nqubits)
+    # theta = 2 * pi + 4 * n * pi
+    elif (theta / (2 * np.pi) - 1) % 2 == 0:
+        return CRZ(symplectic_matrix, control_q, target_q, nqubits, theta)
+    # theta = 3 * pi + 4 * n * pi
+    elif (theta / np.pi - 3) % 4 == 0:
+        symplectic_matrix = CZ(symplectic_matrix, control_q, target_q, nqubits)
+        symplectic_matrix = Z(symplectic_matrix, target_q, nqubits)
+        symplectic_matrix = CNOT(symplectic_matrix, control_q, target_q, nqubits)
+        return Z(symplectic_matrix, target_q, nqubits)
+
+
+def ECR(symplectic_matrix, control_q, target_q, nqubits):
+    symplectic_matrix = S(symplectic_matrix, control_q, nqubits)
+    symplectic_matrix = SX(symplectic_matrix, target_q, nqubits)
+    symplectic_matrix = CNOT(symplectic_matrix, control_q, target_q, nqubits)
+    return X(symplectic_matrix, control_q, nqubits)
+
+
+def _exponent(x1, z1, x2, z2):
+    """Helper function that computes the exponent to which i is raised for the product of the x and z paulis encoded in the symplectic matrix. This is used in _rowsum. The computation is performed parallely over the separated paulis x1[i], z1[i], x2[i] and z2[i].
+
+    Args:
+        x1 (np.array): Bits of the first x paulis.
+        z1 (np.array): Bits of the first z paulis.
+        x2 (np.array): Bits of the second x paulis.
+        z2 (np.array): Bits of the second z paulis.
+
+    Returns:
+        (np.array): The calculated exponents.
+    """
+    exp = np.zeros(x1.shape, dtype=int)
+    x1_eq_z1 = (x1 ^ z1) == False
+    x1_neq_z1 = ~x1_eq_z1
+    x1_eq_0 = x1 == False
+    x1_eq_1 = ~x1_eq_0
+    ind2 = x1_eq_z1 & x1_eq_1
+    ind3 = x1_eq_1 & x1_neq_z1
+    ind4 = x1_eq_0 & x1_neq_z1
+    exp[ind2] = z2[ind2].astype(int) - x2[ind2].astype(int)
+    exp[ind3] = z2[ind3].astype(int) * (2 * x2[ind3].astype(int) - 1)
+    exp[ind4] = x2[ind4].astype(int) * (1 - 2 * z2[ind4].astype(int))
+    return exp
```

Comment:
Not sure why he did not simplify the expression in the paper, but you could write it in a single one, without any `if` required (it's just boolean algebra, with 4 rather than 2 variables):
```py
-2*x1*x2*z1 + 2*x1*x2*z2 + 2*x1*z1*z2 - x1*z2 - 2*x2*z1*z2 + x2*z1
```

Reply:
Btw, it's interesting to notice that is actually a determinant:
```py
[ins] In [1]: import sympy
                                                                                                        
[ins] In [2]: x1, z1, x2, z2 = sympy.symbols("x1 z1 x2 z2")
                                                                                                        
[ins] In [3]: m = sympy.Matrix([[x1, z1, 2 * x1], [x2, z2, 2 * z2], [x2, z1, 1]])
                                                                                                        
[ins] In [4]: m
Out[4]:
Matrix([
[x1, z1, 2*x1],
[x2, z2, 2*z2],
[x2, z1,    1]])
                                                                                                        
[ins] In [5]: for p in [(0, 0), (1, 1), (1, 0), (0, 1)]:
         ...:     t = (-m.det()).subs(zip((x1, z1), p)).simplify()
         ...:     print(f"{p}: {t}")
         ...:
(0, 0): 0
(1, 1): -x2 + z2
(1, 0): z2*(2*x2 - 1)
(0, 1): x2*(1 - 2*z2)
                                                                                                        
[ins] In [6]: -m.det()
Out[6]: -2*x1*x2*z1 + 2*x1*x2*z2 + 2*x1*z1*z2 - x1*z2 - 2*x2*z1*z2 + x2*z1
```

But I'm not sure it's an advantage wrt to just using the function above.

Reply:
(in principle the determinant could be more optimal, if you start from the matrix, because it could cache and reuse part of the computation - but here it's also simpler than an arbitrary determinant, so I'd rather manually collect some of the terms or defining some variables, if the speed-up is soooo relevant...)

Reply:
Interesting, speeding up the `_exponent` function is quite important for measurements actually. I am not sure how to deal with the determinant though, as each `x` and `z` is a matrix itself. I'll probably go with the easier version for now, something like this:
```
x1_x2 = x1 & x2
x1_z2 =  x1 & z2
x2_z1 = x2 & z1
exp = (
-2 * (x1_x2 & z1).astype(int) + 
2 * (x1_x2 & z2).astype(int) + 
2 * (x1_z2 & z1).astype(int) - 
x1_z2.astype(int) - 
2 * (x2_z1 & z2).astype(int)  +
x2_z1.astype(int)
)
```

Reply:
If nothing else, just for the sake of readability, I'd apply the cast immediately to the 4 variables (they are also 4 cast operations instead of 6), and just use `*` instead of `&` (while `|` is only `+ mod 2`, `&` is `*` even in the target set of natural numbers).

Then you can collect as you prefer, if you want to speed up. There is no unique way of doing it, but I'd collect at least the factor of 2, and an arbitrary choice among the equivalent ones (despite not being unique - that's why I wrote it expanded - collecting reduces the number of operations).
E.g.
```py
x1_ = x1.as_type(int)
x2_ = x2.as_type(int)
z1_ = z1.as_type(int)
z2_ = z2.as_type(int)
exp = 2 * (x1_*x2_ * (z2_ - z1_) + z1_*z2_ * (x1_ - x2_)) - x1_*z2_ + x2_*z1_
```

I did this manually (and spontaneously attempting to preserve some symmetry), but you could try to ask Sympy to collect them, and check if you could manage to get even fewer operations in `exp`.

Moreover, I'm appending an `_` to the casted variables, for the general principle of avoiding shadowing. I expect any compiler to be able to drop from memory unused variables, but you could check explicitly if shadowing the name forces the memory free (I don't believe so, but with Python, and even its frameworks, memory management is an art of guessing...).

Reply:
But wait, you cannot cast them before I believe, you want to do the boolean operations first and then sum all the resulting ints. Otherwise you'll end up doing boolean operations with numpy ints that should be considerably slower.

Reply:
As I said, the only boolean operations you're doing before are the `&`, that is the same of the `*` for the domain you're interested in.

Consider that NumPy booleans are still `u8` (they use entire bytes), and that the integer product is an assembly operation (so I'm not user how much you could get out of the `&`...).

For as long as you're not working on bitstrings, I'd suggest to just use `int8` for the type (since you could go negative, but you're bounded to be `|exp| < 10`... actually the bound should be `|exp| <= 2`, but you could get more in the intermediate operations, though it's not relevant).

If you were using bitstrings, then I would agree with you, that could be faster.

Reply:
If you were using bitstrings instead, then I would agree with you, that could be faster. But it's a trade-off with the exceeding casts.

So, eventually you should benchmark them (or print the assembly and read the instructions).

Reply:
I tried running this simple code snippet
```
tottime = 0
for _ in range(1000):
     t = time()
     a * b
     tottime += time() - t
print(tottime/1000)
```
with `a` and `b`:
```
a = np.random.randint(2, size=int(1e6), dtype=int)
b = np.random.randint(2, size=int(1e6), dtype=int)
```
I get on average `8*10^(-4)` with `dtype=int` and `5*10^(-5)` with `dtype=bool`

Reply:
I tested the full function:
- the current one:
```
143 µs ± 1.01 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
- the current one with `np.int8` instead of `int` (that should be an alias for `np.int32` or `np.int64` within NumPy)
```
59 µs ± 264 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
- my proposal
```
20.4 µs ± 27.2 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

(I've been lazy and I did it in the IPython interpreter, I attach the input log [quickbench.py.zip](https://github.com/qiboteam/qibo/files/14219209/quickbench.py.zip), you can just load with `%load` in IPython - it's a bit of a mess, but if you delete a bunch of lines, rearrange, and import `timeit` instead of using the magic it becomes a script in no time)

Reply:
I've retried your script with x and z of shape (100, 1000) (which is a sensible dimension for 1000 qubits) and using `int8` in both cases. I still see my version being faster: `~0.06s` vs `~0.15s` for 1000 repetitions.
 
`quickbench.py`

```
# coding: utf-8
import numpy as np
import timeit

def _exponent(x1, z1, x2, z2):
    x1_x2 = x1 & x2
    x1_z2 = x1 & z2
    x2_z1 = x2 & z1
    return (
        -2 * (x1_x2 & z1).astype(np.int8)
        + 2 * (x1_x2 & z2).astype(np.int8)
        + 2 * (x1_z2 & z1).astype(np.int8)
        - x1_z2.astype(np.int8)
        - 2 * (x2_z1 & z2).astype(np.int8)
        + x2_z1.astype(np.int8)
    )
    
e1 = _exponent
def _exponent(x1, z1, x2, z2):
    x1_ = x1.astype(np.int8)
    x2_ = x2.astype(np.int8)
    z1_ = z1.astype(np.int8)
    z2_ = z2.astype(np.int8)
    return 2 * (x1_*x2_ * (z2_ - z1_) + z1_*z2_ * (x1_ - x2_)) - x1_*z2_ + x2_*z1_
    
e2 = _exponent

n = int(1e3)
nrows = int(1e2)
x1 = np.random.randint(2, size=(nrows,n), dtype=bool)
z1 = np.random.randint(2, size=(nrows,n), dtype=bool)
x2 = np.random.randint(2, size=(nrows,n), dtype=bool)
z2 = np.random.randint(2, size=(nrows,n), dtype=bool)
print(timeit.timeit("e1(x1, z1, x2, z2)", globals=globals(), number=1000))
print(timeit.timeit("e2(x1, z1, x2, z2)", globals=globals(), number=1000))
```

Reply:
I obtained similar numbers, but actually swapped wrt to what you're quoting. I changed once more the script to make the result clearer, here is the report:

```
Benchmarking '6 casts - late'
        mean: 1.2002543163718655        std: 0.015500483036567113
Benchmarking '4 casts - eager'
        mean: 0.8415164417820051        std: 0.01536581466982159

The best is '4 casts - eager'
        confidence: 16.436280886102384sigma
```
[quickbench.py.zip](https://github.com/qiboteam/qibo/files/14221920/quickbench.py.zip)

Reply:
Weird. I get exactly the opposite on my laptop.
```
$ python quickbench.py 
Benchmarking '6 casts - late'
	mean: 0.4699267947999942	std: 0.0028983781006549974
Benchmarking '4 casts - eager'
	mean: 1.423648849600022	std: 0.012184825263011127

The best is '6 casts - late'
	confidence: 76.14670215152795sigma
```
But yeah, on the cluster I get the same as you.

Reply:
I was also thinking that performance wise, maybe we should switch directly to `np.uint8` as the boolean operations are going to be much faster, at least with numpy. Do you know how cupy handles boolean?